File: backend/py/src/crystalith/shared/parsers/csv.py
from __future__ import annotations

import csv
import io

from .interfaces import ParserError
from .types import Chunk


class CSVParser:
    parser_type = "csv"
    supported_mime_types = {"text/csv"}
    supported_extensions = {".csv"}
    page_count: int | None = None

    max_rows_per_chunk = 50
    max_cell_chars = 200

    def parse(self, content: bytes) -> list[Chunk]:
        try:
            text = content.decode("utf-8")
        except UnicodeDecodeError as exc:
            raise ParserError("Unable to decode CSV content as UTF-8") from exc

        if not text.strip():
            return []

        reader = csv.reader(io.StringIO(text))
        rows = list(reader)
        if not rows:
            return []

        raw_header = rows[0]
        header: list[str] = []
        for index, cell in enumerate(raw_header):
            name = (cell or "").strip()
            header.append(name if name else f"col{index + 1}")

        data_rows = rows[1:]
        if not data_rows:
            return []

        chunks: list[Chunk] = []
        for start in range(0, len(data_rows), self.max_rows_per_chunk):
            end = min(len(data_rows), start + self.max_rows_per_chunk)
            block = data_rows[start:end]

            table_lines = _to_markdown_table(header, block, max_cell_chars=self.max_cell_chars)
            metadata = {
                "csv_row_start": start + 1,
                "csv_row_end": end,
            }
            chunks.append(Chunk(text="\n".join(table_lines), metadata=metadata))

        return chunks


def _normalize_cell(value: str, *, max_cell_chars: int) -> str:
    normalized = (value or "").replace("\r\n", "\n").replace("\r", "\n").strip()
    normalized = normalized.replace("|", "\\|")
    normalized = normalized.replace("\n", " ")
    if len(normalized) > max_cell_chars:
        return normalized[: max(0, max_cell_chars - 1)] + "…"
    return normalized


def _to_markdown_table(header: list[str], rows: list[list[str]], *, max_cell_chars: int) -> list[str]:
    column_count = max(1, len(header))

    def normalize_row(row: list[str]) -> list[str]:
        cells: list[str] = []
        for index in range(column_count):
            raw = row[index] if index < len(row) else ""
            cells.append(_normalize_cell(raw, max_cell_chars=max_cell_chars))
        return cells

    header_cells = normalize_row(header)
    lines = [
        "| " + " | ".join(header_cells) + " |",
        "| " + " | ".join(["---"] * column_count) + " |",
    ]
    for row in rows:
        lines.append("| " + " | ".join(normalize_row(row)) + " |")
    return lines

File: apps/server/src/features/sources/parsers/csv.ts
// CSV parser — produces markdown-table chunks with row metadata.
//
// c46: ports v1 parsers/csv.py — 50 rows/chunk, 200 char cell truncation,
// csv_row_start/csv_row_end metadata per chunk.
// c46-fix: returns pages[] so the pipeline can create per-chunk rows with
// individual metadata (instead of re-chunking via generic chunkText).
import type { Parser, ParseResult } from '../parser-registry.ts';

const MAX_ROWS_PER_CHUNK = 50;
const MAX_CELL_CHARS = 200;

/** Parse CSV text into rows (handles quoted fields with commas/newlines). */
function parseCsvRows(text: string): string[][] {
  const rows: string[][] = [];
  let currentRow: string[] = [];
  let currentField = '';
  let inQuotes = false;

  for (let i = 0; i < text.length; i++) {
    const char = text[i]!;

    if (inQuotes) {
      if (char === '"') {
        if (text[i + 1] === '"') {
          currentField += '"';
          i++;
        } else {
          inQuotes = false;
        }
      } else {
        currentField += char;
      }
    } else {
      if (char === '"') {
        inQuotes = true;
      } else if (char === ',') {
        currentRow.push(currentField);
        currentField = '';
      } else if (char === '\n' || char === '\r') {
        // Handle \r\n
        if (char === '\r' && text[i + 1] === '\n') i++;
        currentRow.push(currentField);
        currentField = '';
        if (currentRow.length > 0 && currentRow.some((f) => f !== '')) {
          rows.push(currentRow);
        }
        currentRow = [];
      } else {
        currentField += char;
      }
    }
  }
  // Last field
  if (currentField !== '' || currentRow.length > 0) {
    currentRow.push(currentField);
    if (currentRow.some((f) => f !== '')) {
      rows.push(currentRow);
    }
  }

  return rows;
}

/**
 * c53: escape a cell for markdown-table safety (v1 csv.py:61-67).
 * Pipes become \|, newlines become spaces. MUST run before truncation.
 */
function escapeCell(cell: string): string {
  return cell.replaceAll('|', '\\|').replaceAll(/\r?\n/g, ' ');
}

/** c53: truncate a cell to MAX_CELL_CHARS using the … ellipsis (v1 csv.py:66). */
function truncateCell(cell: string): string {
  return cell.length > MAX_CELL_CHARS ? cell.slice(0, MAX_CELL_CHARS) + '…' : cell;
}

/** c53: escape + truncate, in that order (v1 csv.py:61-67). */
function formatCell(cell: string): string {
  return truncateCell(escapeCell(cell));
}

/** Convert rows to a markdown table string. */
function rowsToMarkdownTable(header: string[], rows: string[][]): string {
  const headerLine = `| ${header.map(formatCell).join(' | ')} |`;
  const separator = `| ${header.map(() => '---').join(' | ')} |`;
  const dataLines = rows.map((row) => `| ${row.map((c) => formatCell(c ?? '')).join(' | ')} |`);
  return [headerLine, separator, ...dataLines].join('\n');
}

export const csvParser: Parser = {
  id: 'csv',
  name: 'CSV Parser',
  mimeTypes: ['text/csv'],
  extensions: ['.csv'],
  async parse(buf: Uint8Array, _filename?: string): Promise<ParseResult> {
    const text = new TextDecoder().decode(buf);
    if (!text.trim()) return { text: '', metadata: { parser: 'csv' } };

    const rows = parseCsvRows(text);
    if (rows.length === 0) return { text: '', metadata: { parser: 'csv' } };

    // Normalize header (empty column names → col{N})
    const rawHeader = rows[0]!;
    const header = rawHeader.map((cell, index) => {
      const name = (cell ?? '').trim();
      return name || `col${index + 1}`;
    });

    const dataRows = rows.slice(1);

    // Chunk into groups of MAX_ROWS_PER_CHUNK — each becomes a page with
    // csv_row_start/csv_row_end metadata (v1 csv.py:48-55).
    const pages: { text: string; metadata: Record<string, unknown> }[] = [];
    for (let start = 0; start < dataRows.length; start += MAX_ROWS_PER_CHUNK) {
      const end = Math.min(dataRows.length, start + MAX_ROWS_PER_CHUNK);
      const block = dataRows.slice(start, end);
      const tableText = rowsToMarkdownTable(header, block);
      pages.push({
        text: tableText,
        metadata: {
          csv_row_start: start + 1, // 1-based (v1)
          csv_row_end: end,
        },
      });
    }

    // Also set full text + top-level metadata for backward compat
    return {
      text: pages.map((p) => p.text).join('\n\n'),
      pages,
      metadata: {
        parser: 'csv',
        row_count: dataRows.length,
        column_count: header.length,
        chunk_count: pages.length,
      },
    };
  },
};
